add drop and recompute media model

agnessnowplow opened the pull request:
The media model written in a drop and recompute fashion until the integration details are finalised. It is less likely it will be a standalone model, depending on the player it should either use a web model or mobile model (Roku) as a base.

The schema.yml describes what each model does but as a whole the idea is that all relevant event level fields are collected under 'interactions', which is aggregated to a page view level, which is the basis for play events. There are further downstream models which aggregate this data from different perspectives. The most important model is the 'media_stats', which aggregates the page view level data to a video/audio level with the key video analytics KPIs in mind.  

Please note that currently it only takes test data from the web model, therefore some key fields from Roku events are currently missing which will result in fails for the 'not_null' tests and incorrect / missing calculations for the relevant data in the models. This should fix itself as soon as we have test data available and populate the correct ids that are currently missing.

emielver commented on models/plays_by_pageview.sql:
```diff
@@ -0,0 +1,140 @@
+    coalesce(sum(p.weight_rate * i.duration), 0)::int as play_time_sec_estimated,
+    min(start_tstamp) start_tstamp,
+    sum(case when i.event_type in ('seek', 'seeked') then 1 else 0 end) as seeks,
+    max(case 
```
Not sure I understand the variable naming logic here, why is this `last_percent_progress`? Isn't it more reporting on most progress made?

agnessnowplow replied:
it refers to the last percent progress that was reached during the play within the page view but I can rename it to `max_percent_progress` to make it clearer

emielver commented on models/plays_by_session.sql:
```diff
@@ -0,0 +1,56 @@
+  seeks,
+  play_time_sec,
+  play_time_sec_muted,
+  play_time_min::int,
```
Maybe worth casting using `CAST(play_time_min as {{ dbt_utils.type_int() }} ) as play_time_min`
here and wherever else we're casting using`::`

emielver commented on models/plays_by_user.sql:
```diff
@@ -0,0 +1,70 @@
+
+left join percent_progress pp
+on pp.user_media_id = p.user_media_id
+
```
Remove trailing lines

emielver commented on models/schema.yml:
```diff
@@ -0,0 +1,67 @@
+models:    
+  - name: mp_events
+    description: "This table collects all media player events as a base table"
+    columns:
```
I guess more column documentation will be added? And it might be worth writing your descriptions in the docs folder and then referencing the documentation written in the docs folder here

agnessnowplow replied:
yes, that will be a separate ticket for sure, this was a very initial draft only :)

agnessnowplow replied:
You will see documentation once I push the updated project.

emielver commented on models/schema.yml:
```diff
@@ -0,0 +1,67 @@
+    description: "This table aggregates media player interactions to a pageview level"
+    columns:
+      - name: play_id
+        description: "The primary key of this table that is unique for each media_id and pageview"
```
I think when referencing tables or other columns it makes sense to put them around single quotes for markdown purposes, e.g. for each `media_id` and `page_view_id`

emielver commented on packages.yml:
```diff
@@ -0,0 +1,3 @@
+packages:
```
Should we consider just using 0.8.0 straight away? Any reason why not?

agnessnowplow replied:
good point, changed it now

emielver commented on models/interactions.sql:
```diff
@@ -0,0 +1,195 @@
+    coalesce(y.player_id, me.html_id, r.content_title) as media_id, -- replace content_title with id field when available
+    round(mp.duration) as duration,
+    coalesce(y.player_id, me.html_id, r.content_title) as title, -- replace ids with title fields when available
+    case when me.media_type = 'audio' then 'audio' else 'video' end as media_type,
```
Why not just get this info from `ref('plays_by_pageview')`?

agnessnowplow replied:
we can't, interactions get aggregated into plays_by_pageview so it has to be calculated upstream

emielver reviewed (changes requested):
Great job Agnes, its coming along really well! I think there's a couple of main points to look out for that are just purely cosmetic:
- Standardising indentation across the board
- Making sure files end on a newline
- Adding as when providing aliases for columns

Then there's the replacing of casting from the `::` format to making use of the various `{{ dbt_utils.type_X }}` to ensure that this model is compatible with all of the different databases that are supported.

I'm also not sure if it's worth implementing a fully incremental logic now, but it seems like a waste to continually recompute so many metrics around users and videos, although if that's out of scope then that's fine too. 

I think breaking down the models into different subsections and therefore directories could also help with the organisation of the project, although that's not necessary. 

I think we can also add some more variables into `dbt_project.yml` to parameterise some settings to make it easier to modify.  

Some of my comments/questions are probably quite ignorant since I haven't worked with the media dataset before so feel free to ignore if the questions don't make sense at all, but overall I'm excited to see how this project continues to evolve!

agnessnowplow commented on models/interactions.sql:
```diff
@@ -0,0 +1,195 @@
+    case when me.media_type = 'audio' then 'audio' else 'video' end as media_type,
+    coalesce(y.schema_vendor||'-'||y.schema_name, me.schema_vendor||'-'||me.schema_name, r.schema_vendor||'-'||r.schema_name)  as media_player_type,
+    e.page_referrer,
+    coalesce(e.page_url, r.content_url) as content_url,
```
Web based media events have the page_url coming from the raw events table which shows where the user was at when they interacted with the media. For embedded YouTube videos we also have the original YouTube video link. To keep both I think what would make sense is to keep the page_url which will be null for Roku, and add an additional 'source_url' field which will be the original YouTube video url, the Roku`s 'content_url' or the 'current_src' for HTML5 depending on the player.

agnessnowplow commented on models/interactions.sql:
```diff
@@ -0,0 +1,195 @@
+
+  select
+    media_id,
+    max(duration) as duration
```
I have noticed a few edge cases where the event came without a duration context for some unknown reason hence the  calculation here. Also, in cases it is a fraction not a whole number hence the earlier rounding on top of this. Thought it is safer to make it coherent, especially that this will be used as a base for calculations later on.

bill-warner commented on models/media_stats.sql:
```diff
@@ -0,0 +1,126 @@
+    group by 1
+)
+
+, user_stats as (
```
Not sure how much these add. More generally I am not sure we need `plays_by_user`. It feels like a slightly odd level of aggregation, essentially a complete view of a single video per user. What do you think?

bill-warner commented on models/media_stats.sql:
```diff
@@ -0,0 +1,126 @@
+    
+    from {{ ref("interactions") }}
+
+    where event_type = 'ready'
```
Do you think we should include impressions in the `plays_by_pageview` model? I know strictly speaking it wouldnt be just plays then but just thinking out loud

agnessnowplow replied:
As we agreed on the group discussion, I have changed the logic to fit the impressions into `plays_by_pageview`. This has led me to realise I need to reconsider KPIs for `media_stats` in general. Currently I only have a rough idea which metric needs which logic (or more than one) out of the following 4 levels of filtering: let's say for `unique users` as a base we could take 1) all interactions including those cases when there is only a `ready` event within the page view so no actual interaction with the video 2) all page views where some interaction with the video happened but not necessarily play event 3) interactions within page views with SOME play_time_sec 4) interactions with valid plays (currently longer than 30 min). You will see I opted for calculating both 3) `unique_users_with_plays` and 4) `unique_users_with_valid_plays` whereas when I look at completion_rate, I only consider option 3) (`completion_rate_by_play`)

bill-warner commented on dbt_project.yml:
```diff
@@ -0,0 +1,27 @@
+  +materialized: view
+
+vars:
+  snowplow__percent_progress_boundaries: {15: 0.15, 25: 0.1, 50: 0.25, 75: 0.25, 100: 0.25}
```
It would be nice to derive the weights from the boundaries meaning the user would only need to supply a list of boundaries

agnessnowplow replied:
In the next commit you will see that the user will only have to fill out a list instead of a dictionary. The weights are automatically calculated within `pivot_base`. One thing to note, assuming that the users do not track `100%` as there is already a separate `ended` event and it is changed to `100%` within the model, I added an if condition to add 100 to it, in case they do not specify it so that the list will be complete. I don't suppose they would not want to see the full picture with the same format. The other option is to not modify the ended events or anything here but keep a separate `ended` field which might be simpler but less coherent for downstream models)

bill-warner reviewed (commented):
Overall I think this is a really good effort. Well done.

I have left plenty of comments throughout but here are some points I would like to highlight:

* The play time calculations I am still a little unclear on. I suggest we work through the comments, make any changes/clarifications and then revisit this part.
* I think we need to think through the session and user level tables. Do we include these in the package as standalone tables, do we provide snippets as to how you might add these cols to the corresponding table in the web package, do we provide models that do this for you. I am not sure at this point. I think we probably need to experiment with that a little. I think what we have already at a sessions and users level is really useful, I think we just need to think how we surface this.
* How do we integrate Roku. It is tricky because Roku's `session_id` is taken from `com_snowplowanalytics_snowplow_client_session_1`, where as all web traffic's session id is `domain_sessionid`. As a result, the mobile base module will surface all Roku events, where as the web base module will surface the remaining media events. How we handle that I am not sure at this stage. It might be that we just focus on either or for now.

Once you have had some time to digest the comments and make any required changes, I suggest we sit down as a 3 and discuss any outstanding questions and think through the points above

agnessnowplow commented on models/plays_by_pageview.sql:
```diff
@@ -0,0 +1,140 @@
+    i.dvce_type,
+    i.os_name,
+    i.os_timezone,
+    coalesce(sum(p.weight_rate * i.duration), 0)::int as play_time_sec_estimated,
```
Yes, I kept it for testing / comparison purposes but it might be an unnecessary field in the end

agnessnowplow commented on models/plays_by_pageview.sql:
```diff
@@ -0,0 +1,140 @@
+    sum(case when i.event_type in ('seek', 'seeked') then 1 else 0 end) as seeks,
+    max(case 
+          when i.event_type = 'ended' then 100 
+          when i.percent_progress = 0 then NULL 
```
It is changed due the distinct count of `percent_progress_reached` calculation within plays_by_user as it seemed superfluous, but after all it is indeed upon the user to decide which percent_progress they would like to track.

agnessnowplow commented on models/plays_by_pageview.sql:
```diff
@@ -0,0 +1,140 @@
+    sum(i.play_time_sec_amended) as play_time_sec,
+    sum(case when i.muted then i.play_time_sec_amended else 0 end) as play_time_sec_muted,
+    coalesce(sum(i.playback_rate * i.play_time_sec) / nullif(sum(i.play_time_sec), 0), 0) as avg_playback_rate,
+    coalesce(sum(i.play_time_sec_amended)/ nullif(max(i.duration), 0), 0) as retention_rate
```
Yes, this is one of the confusing fields. Usually `retention_rate` is viewed across time in the video to see how much the audience is still engaging with the video at what stage but the overall figure either means the current definition OR the % of viewers watching the entire video as you pointed out, which I called `completion_rate`. Once I add the definitions it should be clearer but if it is confusing we can of course rename it to something more descriptive.

agnessnowplow commented on models/plays_by_pageview.sql:
```diff
@@ -0,0 +1,140 @@
+    os_name,
+    os_timezone
+
+  having sum(case when i.event_type in ('play', 'playing') then 1 else 0 end) > 0
```
It should have probably been called `plays` instead of `views` as I only meant to count the number of times the users actually played a video when visiting the page instead of the number of page views with a video embedded.

agnessnowplow commented on models/plays_by_session.sql:
```diff
@@ -0,0 +1,56 @@
+  select
+    domain_sessionid,
+    domain_userid,
+    count(distinct case when media_type = 'video' and is_valid_play then media_id end) as videos_played,
```
Yes, for HTML5 and YouTube we should have in the latest version, just not in the test data yet. For Roku I have not asked for it to be changed but I believe it is only relevant for websites anyway.

agnessnowplow commented on models/interactions.sql:
```diff
@@ -0,0 +1,195 @@
+{{ 
```
initially it is for redhift/postgres but for sure we can add the additional configs to support the rest

agnessnowplow commented on models/interactions.sql:
```diff
@@ -0,0 +1,195 @@
+    mp.current_time as media_start_pos,
+    isnull(mp.playback_rate, 1) as playback_rate,
+    coalesce(y.playback_quality, ve.video_height||'x'||ve.video_width, rd.display_height||'x'||rd.display_width) as playback_quality,
+    case when event_type = 'ended' then 100 else mp.percent_progress end percent_progress,
```
It is up to the user to define which % progress they would like to track, however, as there is a separate `ended` event, they might not add `100` to that list as it would be superfluous to have an extra event just for that (the default values are 15, 25, 50 and 75), that is why for now I opted for updating the percent progress field in those cases to 100%. I'm expecting this to be the only 'correction' we would consider making even though there are many different event types.

agnessnowplow commented on models/interactions.sql:
```diff
@@ -0,0 +1,195 @@
+    e.domain_userid,
+    coalesce(y.player_id, me.html_id, r.content_title) as media_id, -- replace content_title with id field when available
+    round(mp.duration) as duration,
+    coalesce(y.player_id, me.html_id, r.content_title) as title, -- replace ids with title fields when available
```
for Roku we have a content_id and a content_title as well. As we just decided to split Roku from the web models, I will remove this (which was intended as a placeholder for now) and keep the id only

agnessnowplow commented on models/interactions.sql:
```diff
@@ -0,0 +1,195 @@
+		f.event_id,
+    f.event_type,
+    f.start_tstamp,
+		case when f.end_tstamp is null then p.start_tstamp else f.end_tstamp end end_tstamp,
```
I tried to highlight it in the comment but essentially part2 was just a placeholder to eventually calculate the end timestamps of events that are not a `paused` or `ended` event and take them from the web model`'s pageview table`'s `end_tstamp`s to get a more accurate view. Not sure if it is worth the extra calculation and dependence though.

bill-warner replied:
Yeh I guess you could consider either taking the `end_tstamp` from the page views table or use the page ping events to calculate the `end_tstamp` within this model

agnessnowplow replied:
I left it with the simplified version, meaning I will equal the `end_timestamp` of the last event as the `start_tstamp` because as I found that when I added the `end_tstamps` from the `page view` table, the values I got was less than the `start_tstamp` of the last event.

agnessnowplow commented on models/interactions.sql:
```diff
@@ -0,0 +1,195 @@
+	
+	from find_nulls_round_1 f
+
+	inner join prep p
```
It's best to carry the relevant fields through as I can't see any other reason to do it this way with a fresh look now.

agnessnowplow commented on models/interactions.sql:
```diff
@@ -0,0 +1,195 @@
+		event_type,
+		start_tstamp,
+		media_start_pos,
+		lead(start_tstamp, 1) over(partition by page_view_id, media_id order by start_tstamp) end_tstamp,
```
Thanks for your explanation, it makes sense. In the reworked model it will be added to `interactions` already.

bill-warner replied:
Sorry I guess my point here was that it might be cleaner to have the surrogate key defined earlier within the interactions model so it is more obvious what the partition actually represents.

agnessnowplow replied:
I wanted to do that of course but the issue I faced was {{ dbt_utils.surrogate_key(['p.page_view_id', coalesce(y.player_id, me.html_id) ]) }} play_id doesn't work. Do you have a workaround for this perhaps? The only thing I can think of is to add another CTE after prep and before the partitioning CTE just for the surrogate key calculation.

bill-warner replied:
I think you just need to wrap the coalesce in quotes i.e. `{{ dbt_utils.surrogate_key(['p.page_view_id', 'coalesce(y.player_id, me.html_id)' ]) }} play_id`.

agnessnowplow commented on models/plays_by_session.sql:
```diff
@@ -0,0 +1,56 @@
+    sum(seeks) seeks,
+    sum(play_time_sec) as play_time_sec,
+    sum(play_time_sec_muted) as play_time_sec_muted,
+    sum(play_time_sec / 60) play_time_min,
```
sure, minutes are removed in the new commit (for muted too)

agnessnowplow commented on models/plays_by_user.sql:
```diff
@@ -0,0 +1,70 @@
+    sum(seeks) as seeks,
+    sum(case when is_complete_play then 1 else 0 end) as complete_plays
+  
+  from {{ ref("plays_by_pageview") }}
```
We agreed to leave out plays_by_user so this is obsolete now.

agnessnowplow commented on models/mp_events.sql:
```diff
@@ -0,0 +1,53 @@
+  left join {{ source('atomic', 'com_snowplowanalytics_snowplow_web_page_1') }} as pvc
+  on pvc.root_id = e.event_id
+  
+  where e.collector_tstamp >= '2022-01-18' and e.event_name = 'media_player_event'
```
added `snowplow__mp_start_date` to the project.yml file as a new variable

agnessnowplow commented on models/plays_by_pageview.sql:
```diff
@@ -0,0 +1,140 @@
+        column='percent_progress',
+        values=dbt_utils.get_column_values( table=ref('pivot_base'), column='percent_progress', default=[]) | sort,
+        alias=True,
+        agg='count',
```
I changed it for the current model as it is indeed simpler but the drawback is that it is not cross db compatible this way. I wonder if we can introduce a db variable and use it to change the function automatically without having to save a separate model for Snowflake or BigQuery. Unless there will be a lot of things to change in the end but it doesn't look like it at this stage.

agnessnowplow replied:
Just reading about this: https://docs.getdbt.com/reference/dbt-jinja-functions/dispatch, perhaps we can apply something similar in this case.

agnessnowplow commented on models/plays_by_session.sql:
```diff
@@ -0,0 +1,56 @@
+    sum(play_time_sec_muted) as play_time_sec_muted,
+    sum(play_time_sec / 60) play_time_min,
+    sum(play_time_sec_muted / 60) as play_time_min_muted,
+    sum(play_time_sec) / sum(duration) as avg_retention_rate,
```
you are right, I changed it to average based on the individual level calculation

agnessnowplow commented on models/plays_by_pageview.sql:
```diff
@@ -0,0 +1,140 @@
+    max(i.end_tstamp) end_tstamp,
+    sum(i.play_time_sec_amended) as play_time_sec,
+    sum(case when i.muted then i.play_time_sec_amended else 0 end) as play_time_sec_muted,
+    coalesce(sum(i.playback_rate * i.play_time_sec) / nullif(sum(i.play_time_sec), 0), 0) as avg_playback_rate,
```
You are right, let's leave them as nulls in case there is no play time. I also removed the coalesce from the subsequent line, the retention rate, as there should always be a duration and if not, the null value will indicate there is a data issue.

agnessnowplow commented on models/plays_by_pageview.sql:
```diff
@@ -0,0 +1,140 @@
+    {{ dbt_utils.hash('i.page_view_id||i.media_id') }} play_id,
+    i.page_view_id,
+    i.media_id,
+    max(i.domain_sessionid) as domain_sessionid,
```
I have found three cases (out of 70) where the same page_view_id has multiple session_ids. All coming from the YouTube data. In one case it looks like the person got on and off the VPN, in the others it looks like the user was on pause for an hour or more and then continued playing.

agnessnowplow replied:
I removed the max, deducing below the prep CTE should take care of these edge cases, the start_tstamp of the first event within the page_view will take precedence (following a similar logic to the web model)

agnessnowplow commented:
Thank you both again for the reviews, I have reworked the model taking into consideration the recommendations you suggested. I have removed Roku from the model for now so that it is cleaner and the test don't fail anymore. I have also added documentation for the columns so that we can be clear on the definitions behind the KPIs. There are still a handful of points that might still be open that needs further discussion, I tried to reply to comments everywhere it is needed, feel free to resolve any open points you raised to get a clearer view of what is actually left to discuss / rework.

bill-warner commented:
Thanks for making this changes. I have started to review the code base again but stopped because I realised it doesn't actually run e2e. There is 1 issue in the `mp_events` model where you filter on `page_view_id` not `pvc.id`. The remaining issues I am unsure on. Apologies in advance if it is me misconfiguring the model but I have left the configs as is. Once it is running I will have another look :)

agnessnowplow commented:
Apologies, the join is now fixed. The rest of the models run fine.

bill-warner commented:
I am still getting an error. Any idea why that is?

<img width="1022" alt="image" src="https://user-images.githubusercontent.com/71764352/156373635-4f66d0fc-f4a7-4696-ad58-e4855934738e.png">

agnessnowplow commented:
I was just about to comment something similar but I'm glad it works now! I could only reproduce your issue if I first run the pivot_base without 10 within `snowplow__percent_progress_boundaries` and then try to run plays by page view with a different percent progress configuration list where 10 is added to the list within `snowplow__percent_progress_boundaries`. They are indeed dependent.

agnessnowplow commented on models/interactions.sql:
```diff
@@ -0,0 +1,195 @@
+on p.event_id = c.event_id
+
+left join corrected_duration d
+on d.media_id = p.media_id
```
I hope this is fixed now, I'm not aware of any surefire way to check this in vs code at least, as it looked fine on my side previously too. After poking around a bit I found a setting `files: Eol` where the default end of line character was set to `auto`, which according to online forums can cause these issues for git. I changed that to `\n `but I guess it doesn't get applied retrospectively.

bill-warner reviewed (commented):
I like the changes. Good work. I think this is shaping up nicely.

I am still slightly unsure about the play time calculation. I think it is less to do with your logic, more with the underlying data. I have pulled out some rows of data which I had questions on [here](https://docs.google.com/spreadsheets/d/1BTN2u13JG7wt3bWt4oNGqoEYr7nVGR2wEgEXScUTNQ4/edit?usp=sharing). I have left comments in the relevant cells.

More generally on this time calculation it seems like we have two approaches to this. I have outlined these below and my initial thoughts on the two. I am not wedded to either method yet but I would be interested to get both your thoughts on this. 

### 1. Using tstamp of media events (Current implementation)

#### Pros:

* Can calc abs time between events.
* Assuming clean data, should be the most accurate calc.

#### Cons:

* Relatively expensive to use lag/lead functions across event level data
* The user must be tracking a specific set of events for this to work i.e. play, pause, seek, ended. Unsure how tracking additional media events will affect this calc.
* Reliant on accurate tstamps and sequence of events being correct.
* If there is only a single play event, you wont have any watch time. Could use page pings to accommodate this fact.
* Complex logic, difficult to follow assumptions e.g. how seek events are being treated, what happens if elapsed time > play time


### 2. Use more granular percentage boundaries (e.g. every 10%)

#### Pros:

* Less prone to significant error (I think)
* Cheaper to calc
* Would allow for more granular retention curves.
* Only need to capture boundary events for this calc to work.
* Accuracy can be increased by increasing number of boundaries. 

#### Cons:

* Purely an estimate of play time, rather than abs time. 
* Increases number of percentage boundary events
* Hard to calc things that require fine grain measurements such as `is_valid_play` which could be `play_time_sec > 1s`

I haven't gone into the minutiae on this review as I feel the decision we make on the above point will impact the design a bit.

agnessnowplow commented on models/plays_by_pageview.sql:
```diff
@@ -120,21 +87,24 @@ select
   d.play_time_sec_muted,
   d.play_time_sec_estimated,
-  d.retention_rate,
+  case when d.play_time_sec > 0 then true else false end is_played,
```
sure, we could do a count of plays as the basis of the logic to be 100% correct

agnessnowplow commented on models/plays_by_pageview.sql:
```diff
@@ -10,91 +9,59 @@
-    coalesce(sum(i.play_time_sec_amended)/ nullif(max(i.duration), 0), 0) as retention_rate
-  
+    sum(case when i.is_muted then i.play_time_sec_amended else 0 end) as play_time_sec_muted,
+    sum(i.playback_rate * i.play_time_sec) / nullif(sum(i.play_time_sec), 0) as avg_playback_rate,
```
`play_time_sec_amended` is renamed to `play_time_sec` in downstream models. It is probably cleaner if the `play_time_sec` in this model is called `play_time_sec_original` and call `play_time_sec_amended` `play_time_sec` right away.

agnessnowplow replied:
This is not relevant anymore. I have only kept one play_time_sec calculation which is the estimate based on the percent progresses reached, as discussed earlier.

agnessnowplow commented on models/media_stats.sql:
```diff
@@ -1,126 +1,131 @@
-    from {{ ref("plays_by_pageview") }}
+  select
+    media_id,
+    min(start_tstamp) as first_play,
```
I think both could be useful, especially now that we decided to reconsider the playtime calculation so that valid plays will become less accurate in certain scenarios

agnessnowplow commented on models/media_stats.sql:
```diff
@@ -1,126 +1,131 @@
+    min(start_tstamp) as first_play,
+    max(start_tstamp) as last_play,
+    count(*) as valid_plays,
+    count(distinct domain_userid) as unique_users_with_valid_plays,
```
In the end I decided to leave this out and only make a distinction bw `users` (who clicked play) and users with complete plays as I find it slightly superfluous, but shortened all the key metrics with `with` to `w`

agnessnowplow commented on models/pivot_base.sql:
```diff
@@ -1,26 +1,43 @@
-    
-    {% if not loop.last %}
+{% for element in var("snowplow__percent_progress_boundaries") %}
+  {% if element < 0 or element > 100 %}
```
this is done now, thanks for the suggestion.

agnessnowplow commented:
I have reworked the model according to the less expensive and less error prone `play_time_sec` calculation, and that is based on the percentage progresses reached. The more the users track, the more accurate the estimates should be. This simplified the model quite a bit after all, but it also meant that I reconsidered the KPIs (some I renamed, some I removed, some I changed the logic for) I have changed the logic for `avg_retention_rate` which will be the average of the latest percentage progress reached before any seek events to see where the interest drops. This is different from the 'avg_percentage_played' which is the total play_time_sec (including rewatches) divided by the media duration.

emielver commented:
Closed since we have the incremental model instead.
